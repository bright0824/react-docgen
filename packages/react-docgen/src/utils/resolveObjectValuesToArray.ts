import resolveToValue from './resolveToValue.js';
import type { NodePath } from '@babel/traverse';
import type {
  CallExpression,
  Identifier,
  NumericLiteral,
  StringLiteral,
} from '@babel/types';

function isObjectValuesCall(path: NodePath): path is NodePath<CallExpression> {
  if (!path.isCallExpression() || path.node.arguments.length !== 1) {
    return false;
  }

  const callee = path.get('callee');

  if (!callee.isMemberExpression()) {
    return false;
  }
  const object = callee.get('object');
  const property = callee.get('property');

  return (
    object.isIdentifier({ name: 'Object' }) &&
    property.isIdentifier({ name: 'values' })
  );
}

// Resolves an ObjectExpression or an ObjectTypeAnnotation
function resolveObjectToPropMap(object: NodePath): Map<string, string> | null {
  if (object.isObjectExpression()) {
    const values = new Map<string, string>();
    let error = false;

    object.get('properties').forEach((propPath) => {
      if (error || propPath.isObjectMethod()) return;

      if (propPath.isObjectProperty()) {
        const key = propPath.get('key') as NodePath<
          Identifier | NumericLiteral | StringLiteral
        >;
        let name: string;

        // Key is either Identifier or Literal
        if (key.isIdentifier()) {
          name = key.node.name;
        } else if (key.isNumericLiteral() || key.isStringLiteral()) {
          name = `${key.node.value}`;
        } else {
          error = true;

          return;
        }

        const valuePath = resolveToValue(propPath.get('value'));
        const value = valuePath.isStringLiteral()
          ? `"${valuePath.node.value}"`
          : valuePath.isNumericLiteral()
          ? `${valuePath.node.value}`
          : // we return null here because there are a lot of cases and we don't know yet what we need to handle
            'null';

        values.set(name, value);
      } else if (propPath.isSpreadElement()) {
        const spreadObject = resolveToValue(propPath.get('argument'));
        const spreadValues = resolveObjectToPropMap(spreadObject);

        if (!spreadValues) {
          error = true;

          return;
        }
        for (const entry of spreadValues.entries()) {
          const [key, value] = entry;

          values.set(key, value);
        }
      }
    });

    if (!error) {
      return values;
    }
  }

  return null;
}

/**
 * Returns an ArrayExpression which contains all the values resolved from an object
 *
 * Ignores setters in objects
 *
 * Returns null in case of
 *  unresolvable spreads
 *  computed identifier values
 */
export default function resolveObjectValuesToArray(
  path: NodePath,
): string[] | null {
  if (isObjectValuesCall(path)) {
    const argument = path.get('arguments')[0];
    const objectExpression = resolveToValue(
      // isObjectValuesCall already asserts that there is at least one argument, hence the non-null-assertion
      argument!,
    );
    const values = resolveObjectToPropMap(objectExpression);

    if (values) {
      return Array.from(values.values());
    }
  }

  return null;
}
