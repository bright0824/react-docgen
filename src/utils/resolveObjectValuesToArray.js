/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import recast from 'recast';
import resolveToValue from './resolveToValue';

type ObjectPropMap = {
  properties: Array<string>,
  values: Object,
};

const {
  types: { ASTNode, NodePath, builders, namedTypes: types },
} = recast;

function isObjectValuesCall(node: ASTNode): boolean {
  return (
    types.CallExpression.check(node) &&
    node.arguments.length === 1 &&
    types.MemberExpression.check(node.callee) &&
    types.Identifier.check(node.callee.object) &&
    node.callee.object.name === 'Object' &&
    types.Identifier.check(node.callee.property) &&
    node.callee.property.name === 'values'
  );
}

function isWhitelistedObjectProperty(prop) {
  return (
    (types.Property.check(prop) &&
      ((types.Identifier.check(prop.key) && !prop.computed) ||
        types.Literal.check(prop.key))) ||
    types.SpreadElement.check(prop)
  );
}

function isWhiteListedObjectTypeProperty(prop) {
  return (
    types.ObjectTypeProperty.check(prop) ||
    types.ObjectTypeSpreadProperty.check(prop) ||
    types.TSPropertySignature.check(prop)
  );
}

// Resolves an ObjectExpression or an ObjectTypeAnnotation
export function resolveObjectToPropMap(
  object: NodePath,
  raw: boolean = false,
): ?ObjectPropMap {
  if (
    (types.ObjectExpression.check(object.value) &&
      object.value.properties.every(isWhitelistedObjectProperty)) ||
    (types.ObjectTypeAnnotation.check(object.value) &&
      object.value.properties.every(isWhiteListedObjectTypeProperty)) ||
    (types.TSTypeLiteral.check(object.value) &&
      object.value.members.every(isWhiteListedObjectTypeProperty))
  ) {
    const properties = [];
    let values = {};
    let error = false;
    const members = types.TSTypeLiteral.check(object.value)
      ? object.get('members')
      : object.get('properties');
    members.each(propPath => {
      if (error) return;
      const prop = propPath.value;

      if (prop.kind === 'get' || prop.kind === 'set') return;

      if (
        types.Property.check(prop) ||
        types.ObjectTypeProperty.check(prop) ||
        types.TSPropertySignature.check(prop)
      ) {
        // Key is either Identifier or Literal
        const name = prop.key.name || (raw ? prop.key.raw : prop.key.value);
        const propValue = propPath.get(name).parentPath.value;
        const value =
          propValue.value.value ||
          (raw ? propValue.value.raw : propValue.value.value);

        if (properties.indexOf(name) === -1) {
          properties.push(name);
        }
        values[name] = value;
      } else if (
        types.SpreadElement.check(prop) ||
        types.ObjectTypeSpreadProperty.check(prop)
      ) {
        let spreadObject = resolveToValue(propPath.get('argument'));
        if (types.GenericTypeAnnotation.check(spreadObject.value)) {
          const typeAlias = resolveToValue(spreadObject.get('id'));
          if (types.ObjectTypeAnnotation.check(typeAlias.get('right').value)) {
            spreadObject = resolveToValue(typeAlias.get('right'));
          }
        }

        const spreadValues = resolveObjectToPropMap(spreadObject);
        if (!spreadValues) {
          error = true;
          return;
        }
        spreadValues.properties.forEach(spreadProp => {
          if (properties.indexOf(spreadProp) === -1) {
            properties.push(spreadProp);
          }
        });
        values = { ...values, ...spreadValues.values };
      }
    });

    if (!error) {
      return { properties: properties.sort(), values };
    }
  }

  return null;
}

/**
 * Returns an ArrayExpression which contains all the keys resolved from an object
 *
 * Ignores setters in objects
 *
 * Returns null in case of
 *  unresolvable spreads
 *  computed identifier values
 */
export default function resolveObjectValuesToArray(path: NodePath): ?NodePath {
  const node = path.node;

  if (isObjectValuesCall(node)) {
    const objectExpression = resolveToValue(path.get('arguments').get(0));
    const propMap = resolveObjectToPropMap(objectExpression);

    if (propMap) {
      const nodes = propMap.properties.map(prop => {
        const value = propMap.values[prop];

        return typeof value === 'undefined'
          ? builders.literal(null)
          : builders.literal(value);
      });

      return new NodePath(builders.arrayExpression(nodes));
    }
  }

  return null;
}
