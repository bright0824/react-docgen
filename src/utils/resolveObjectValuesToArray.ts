import {
  ASTNode,
  NodePath as NodePathConstructor,
  builders,
  namedTypes as t,
} from 'ast-types';
import resolveToValue from './resolveToValue';
import type { Importer } from '../parse';
import type { NodePath } from 'ast-types/lib/node-path';
import { LiteralBuilder } from 'ast-types/gen/builders';

interface ObjectPropMap {
  properties: string[];
  values: Record<string, unknown>;
}

function isObjectValuesCall(node: ASTNode): boolean {
  return (
    t.CallExpression.check(node) &&
    node.arguments.length === 1 &&
    t.MemberExpression.check(node.callee) &&
    t.Identifier.check(node.callee.object) &&
    node.callee.object.name === 'Object' &&
    t.Identifier.check(node.callee.property) &&
    node.callee.property.name === 'values'
  );
}

function isWhitelistedObjectProperty(prop) {
  return (
    (t.Property.check(prop) &&
      ((t.Identifier.check(prop.key) && !prop.computed) ||
        t.Literal.check(prop.key))) ||
    t.SpreadElement.check(prop)
  );
}

function isWhiteListedObjectTypeProperty(prop) {
  return (
    t.ObjectTypeProperty.check(prop) ||
    t.ObjectTypeSpreadProperty.check(prop) ||
    t.TSPropertySignature.check(prop)
  );
}

// Resolves an ObjectExpression or an ObjectTypeAnnotation
export function resolveObjectToPropMap(
  object: NodePath,
  importer: Importer,
  raw = false,
): ObjectPropMap | null {
  if (
    (t.ObjectExpression.check(object.value) &&
      object.value.properties.every(isWhitelistedObjectProperty)) ||
    (t.ObjectTypeAnnotation.check(object.value) &&
      object.value.properties.every(isWhiteListedObjectTypeProperty)) ||
    (t.TSTypeLiteral.check(object.value) &&
      object.value.members.every(isWhiteListedObjectTypeProperty))
  ) {
    const properties: string[] = [];
    let values = {};
    let error = false;
    const members = t.TSTypeLiteral.check(object.value)
      ? object.get('members')
      : object.get('properties');
    members.each(propPath => {
      if (error) return;
      const prop = propPath.value;

      if (prop.kind === 'get' || prop.kind === 'set') return;

      if (
        t.Property.check(prop) ||
        t.ObjectTypeProperty.check(prop) ||
        t.TSPropertySignature.check(prop)
      ) {
        // Key is either Identifier or Literal
        // @ts-ignore
        const name = prop.key.name || (raw ? prop.key.raw : prop.key.value);
        const propValue = propPath.get(name).parentPath.value;
        const value =
          propValue.value.value ||
          (raw ? propValue.value.raw : propValue.value.value);

        if (properties.indexOf(name) === -1) {
          properties.push(name);
        }
        values[name] = value;
      } else if (
        t.SpreadElement.check(prop) ||
        t.ObjectTypeSpreadProperty.check(prop)
      ) {
        let spreadObject = resolveToValue(propPath.get('argument'), importer);
        if (t.GenericTypeAnnotation.check(spreadObject.value)) {
          const typeAlias = resolveToValue(spreadObject.get('id'), importer);
          if (t.ObjectTypeAnnotation.check(typeAlias.get('right').value)) {
            spreadObject = resolveToValue(typeAlias.get('right'), importer);
          }
        }

        const spreadValues = resolveObjectToPropMap(spreadObject, importer);
        if (!spreadValues) {
          error = true;
          return;
        }
        spreadValues.properties.forEach(spreadProp => {
          if (properties.indexOf(spreadProp) === -1) {
            properties.push(spreadProp);
          }
        });
        values = { ...values, ...spreadValues.values };
      }
    });

    if (!error) {
      return { properties: properties.sort(), values };
    }
  }

  return null;
}

/**
 * Returns an ArrayExpression which contains all the values resolved from an object
 *
 * Ignores setters in objects
 *
 * Returns null in case of
 *  unresolvable spreads
 *  computed identifier values
 */
export default function resolveObjectValuesToArray(
  path: NodePath,
  importer: Importer,
): NodePath | null {
  const node = path.node;

  if (isObjectValuesCall(node)) {
    const objectExpression = resolveToValue(
      path.get('arguments').get(0),
      importer,
    );
    const propMap = resolveObjectToPropMap(objectExpression, importer);

    if (propMap) {
      const nodes = propMap.properties.map(prop => {
        const value = propMap.values[prop] as Parameters<LiteralBuilder>[0];

        return typeof value === 'undefined'
          ? builders.literal(null)
          : builders.literal(value);
      });

      return new NodePathConstructor(builders.arrayExpression(nodes));
    }
  }

  return null;
}
