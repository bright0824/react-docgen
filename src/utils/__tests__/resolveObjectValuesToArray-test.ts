import {
  statement,
  noopImporter,
  makeMockImporter,
  expressionLast,
} from '../../../tests/utils';
import resolveObjectValuesToArray from '../resolveObjectValuesToArray';

describe('resolveObjectValuesToArray', () => {
  const mockImporter = makeMockImporter({
    foo: statement(`
      export default {
        1: "bar",
        2: "foo",
        3: 0,
        4: 5,
        5: undefined,
        6: null,
        [7]: 7,
        ['foo']: "foo",
      };
    `).get('declaration'),

    bar: statement(`
      export default {
        bar: 'bar',
      };
    `).get('declaration'),
  });

  it('resolves Object.values with strings', () => {
    const path = expressionLast(
      ['var foo = { 1: "bar", 2: "foo" };', 'Object.values(foo);'].join('\n'),
    );

    expect(resolveObjectValuesToArray(path, noopImporter)).toMatchSnapshot();
  });

  it('resolves Object.values with numbers', () => {
    const path = expressionLast(
      ['var foo = { 1: 0, 2: 5 };', 'Object.values(foo);'].join('\n'),
    );

    expect(resolveObjectValuesToArray(path, noopImporter)).toMatchSnapshot();
  });

  it('resolves Object.values with undefined or null', () => {
    const path = expressionLast(
      ['var foo = { 1: null, 2: undefined };', 'Object.values(foo);'].join(
        '\n',
      ),
    );

    expect(resolveObjectValuesToArray(path, noopImporter)).toMatchSnapshot();
  });

  it('resolves Object.values with literals as computed key', () => {
    const path = expressionLast(
      ['var foo = { ["bar"]: 1, [5]: 2};', 'Object.values(foo);'].join('\n'),
    );

    expect(resolveObjectValuesToArray(path, noopImporter)).toMatchSnapshot();
  });

  it('does not resolve Object.values with complex computed key', () => {
    const path = expressionLast(
      ['var foo = { [()=>{}]: 1, [5]: 2};', 'Object.values(foo);'].join('\n'),
    );

    expect(resolveObjectValuesToArray(path, noopImporter)).toBeNull();
  });

  it('resolves Object.values when using resolvable spread', () => {
    const path = expressionLast(
      [
        'var bar = { doo: 4 }',
        'var foo = { boo: 1, foo: 2, ...bar };',
        'Object.values(foo);',
      ].join('\n'),
    );

    expect(resolveObjectValuesToArray(path, noopImporter)).toMatchSnapshot();
  });

  it('resolves Object.values when using getters', () => {
    const path = expressionLast(
      [
        'var foo = { boo: 1, foo: 2, get bar() {} };',
        'Object.values(foo);',
      ].join('\n'),
    );

    expect(resolveObjectValuesToArray(path, noopImporter)).toMatchSnapshot();
  });

  it('resolves Object.values when using setters', () => {
    const path = expressionLast(
      [
        'var foo = { boo: 1, foo: 2, set bar(e) {} };',
        'Object.values(foo);',
      ].join('\n'),
    );

    expect(resolveObjectValuesToArray(path, noopImporter)).toMatchSnapshot();
  });

  it('resolves Object.values when using methods', () => {
    const path = expressionLast(
      ['var foo = { boo: 1, foo: 2, bar(e) {} };', 'Object.values(foo);'].join(
        '\n',
      ),
    );

    expect(resolveObjectValuesToArray(path, noopImporter)).toMatchSnapshot();
  });

  it('resolves Object.values but ignores duplicates', () => {
    const path = expressionLast(
      [
        'var bar = { doo: 4, doo: 5 }',
        'var foo = { boo: 1, foo: 2, doo: 1, ...bar };',
        'Object.values(foo);',
      ].join('\n'),
    );

    expect(resolveObjectValuesToArray(path, noopImporter)).toMatchSnapshot();
  });

  it('resolves Object.values but ignores duplicates with getter and setter', () => {
    const path = expressionLast(
      ['var foo = { get x() {}, set x(a) {} };', 'Object.values(foo);'].join(
        '\n',
      ),
    );

    expect(resolveObjectValuesToArray(path, noopImporter)).toMatchSnapshot();
  });

  it('does not resolve Object.values when using unresolvable spread', () => {
    const path = expressionLast(
      ['var foo = { bar: 1, foo: 2, ...bar };', 'Object.values(foo);'].join(
        '\n',
      ),
    );

    expect(resolveObjectValuesToArray(path, noopImporter)).toBeNull();
  });

  it('can resolve imported objects passed to Object.values', () => {
    const path = expressionLast(`
      import foo from 'foo';
      Object.values(foo);
    `);

    expect(resolveObjectValuesToArray(path, mockImporter)).toMatchSnapshot();
  });

  it('can resolve spreads from imported objects', () => {
    const path = expressionLast(`
      import bar from 'bar';
      var abc = { foo: 'foo', baz: 'baz', ...bar };
      Object.values(abc);
    `);

    expect(resolveObjectValuesToArray(path, mockImporter)).toMatchSnapshot();
  });
});
