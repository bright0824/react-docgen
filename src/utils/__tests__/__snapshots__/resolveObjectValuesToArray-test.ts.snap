// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolveObjectValuesToArray can resolve imported objects passed to Object.values 1`] = `
Array [
  "\\"bar\\"",
  "\\"foo\\"",
  "0",
  "5",
  "null",
  "null",
  "7",
  "\\"foo\\"",
]
`;

exports[`resolveObjectValuesToArray can resolve spreads from imported objects 1`] = `
Array [
  "\\"foo\\"",
  "\\"baz\\"",
  "\\"bar\\"",
]
`;

exports[`resolveObjectValuesToArray resolves Object.values but ignores duplicates 1`] = `
Array [
  "1",
  "2",
  "5",
]
`;

exports[`resolveObjectValuesToArray resolves Object.values but ignores duplicates with getter and setter 1`] = `Array []`;

exports[`resolveObjectValuesToArray resolves Object.values when using getters 1`] = `
Array [
  "1",
  "2",
]
`;

exports[`resolveObjectValuesToArray resolves Object.values when using methods 1`] = `
Array [
  "1",
  "2",
]
`;

exports[`resolveObjectValuesToArray resolves Object.values when using resolvable spread 1`] = `
Array [
  "1",
  "2",
  "4",
]
`;

exports[`resolveObjectValuesToArray resolves Object.values when using setters 1`] = `
Array [
  "1",
  "2",
]
`;

exports[`resolveObjectValuesToArray resolves Object.values with literals as computed key 1`] = `
Array [
  "1",
  "2",
]
`;

exports[`resolveObjectValuesToArray resolves Object.values with numbers 1`] = `
Array [
  "0",
  "5",
]
`;

exports[`resolveObjectValuesToArray resolves Object.values with strings 1`] = `
Array [
  "\\"bar\\"",
  "\\"foo\\"",
]
`;

exports[`resolveObjectValuesToArray resolves Object.values with undefined or null 1`] = `
Array [
  "null",
  "null",
]
`;
